streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -83,9 +83,26 @@ public void register(final StateRestorer restorer) {
 
         final Set<TopicPartition> restoringPartitions = new HashSet<>(needsRestoring.keySet());
         try {
-            final ConsumerRecords<byte[], byte[]> allRecords = poll(restoreConsumer, 10);
-            for (final TopicPartition partition : restoringPartitions) {
-                restorePartition(allRecords, partition, active.restoringTaskFor(partition));
+            final Map<TopicPartition, Long> endOffsets = restoreConsumer.endOffsets(restoringPartitions);
+            while (!needsRestoring.isEmpty()) {
+                final ConsumerRecords<byte[], byte[]> records = poll(restoreConsumer, 10);
+                if (records.count() == 0) {
+                    break;
```

Comment:
Do we need to check if restore is completed for some partitions? I think, with EOS and commit markers, there is a corner case that the check below does not detect that restore is complete even if we fetched all data (but not the final commit marker). For this case, records.count() could be zero but the actual `position()` for a partitions was advanced by 1 to step over the commit marker.

Reply:
I am not too clear on this point. Do you have something specific in mind?

Reply:
I think we need the same check as in https://github.com/apache/kafka/pull/4901/files#diff-46ed6d177221c8778965ecb1b6657be3R101

(might be good to extract this into a private method)

We should cover this corner case with a unit test, too. As the tests pass atm, it seems the corner case is not covered yet.

Reply:
To summarize on what this check does, I have found that if this check is removed (```if records.count()==0```) then hanging tests will result since needsRestoring might still contain partitions even if poll() no longer returns any records. However, when I dug through the older version of the code, I could not find the check you are referring to. My closest guess is that you are thinking about comparing ```endOffsets.get(partition)``` to the restored offsets for that particular partition. (note, it is not the ```updatedEndOffsets``` field which is strictly used only in restore). Is this what you mean? Currently, my understanding is sketchy at best.

Reply:
Sorry for not expressing my thoughts clearly.

The check `if (records.count()==0)` if fine. However, I think just doing a `break` is not good enough. Before the `break` we need to check if restore has completed for any partitions that is in restore phase -- and if yes, remove those partitions from `needsRestore` etc. Otherwise, it might happen that a partition stays in "restoring" phase forever, because `count()==0` is always zero and we never to the check in line 101.

Does this make sense?